compare-metrics.py

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env python3
+
+"""compare-metrics
+Compares the metrics before and after a tree-sitter-language update in order to
+discover whether there are differences.
+
+The metrics are saved as json files, and to compare them, a specific
+json-diff has been adopted.
+
+1. Install npm on your system
+2. Install json-diff from npm running: npm install -g json-diff
+
+Usage: ./compare-metrics.py -p METRICS_DIR_PREFIX
+"""
+
+import argparse
+import pathlib
+import subprocess
+import sys
+import typing as T
+
+# The /tmp directory will be used as workdir
+WORKDIR = pathlib.Path("/tmp")
+
+# Run a subprocess and return the value
+def run_subprocess(cmd: str, *args: T.Any) -> subprocess.CompletedProcess:
+    return subprocess.run(
+        [cmd, *args],
+        text=True,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT,
+    )
+
+
+def main() -> None:
+    parser = argparse.ArgumentParser(
+        description="This tool compares the metrics before and after "
+        "a tree-sitter-language update in order to discover whether "
+        "there are differences.",
+        epilog="The source code of this program can be found on "
+        "GitHub at https://github.com/mozilla/rust-code-analysis",
+    )
+
+    # Arguments
+    parser.add_argument(
+        "-p",
+        "--path",
+        type=str,
+        required=True,
+        help="Prefix of the directories containing the metrics",
+    )
+
+    # Parse arguments
+    args = parser.parse_args()
+
+    # Old metrics directory
+    old_dir = WORKDIR / (args.path + "-old")
+    # New metrics directory
+    new_dir = WORKDIR / (args.path + "-new")
+
+    # Run comparisons
+    old_paths = pathlib.Path(old_dir).glob("**/*.json")
+    new_paths = pathlib.Path(new_dir).glob("**/*.json")
+
+    for old_filename, new_filename in zip(old_paths, new_paths):
+        ret_value = run_subprocess("json-diff", "-C", old_filename, new_filename)
+        if ret_value.returncode != 0:
+            print(ret_value.stdout)
+            input_value = input("Continue comparison (y/n)? ")
+            if input_value == "n":
+                sys.exit(0)
+            print()
```

Comment:
It might be better to show the diffs in an "explorable way".
That is, start with showing the list of directories in which there are differences, let the user navigate within the directories down to the actual files where there are differences, and then down to the functions where there are differences.
This might be easier when there are many differences.

We can do this later though.

Reply:
Now there are two modes:

- The `interactive` mode prints on screen the differences among metrics of each file, asking a user if he wants to continue the comparison
- The `compare` mode dumps the json files of differences in a specific directory and prints on screen the paths of files containing the differences.

Reply:
Removed `interactive` mode because on many files it's pretty cumbersome to use